### PR TITLE
Twelve-days: Add exercise for Twelve Days

### PR DESCRIPTION
**Pull request by Akasurde**
Exercise that outputs the lyrics to 'The Twelve Days of Christmas'

Fixes https://github.com/exercism/xgo/issues/397

Signed-off-by: Abhijeet Kasurde akasurde@redhat.com

### CODE REVIEW

File: exercises/twelve-days/example.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	8:  "eighth",
+	9:  "ninth",
+	10: "tenth",
+	11: "eleventh",
+	12: "twelfth",
+}
+
+func verses(i int) string {
+	var gifts = ""
+	for frm := i; frm > 0; frm-- {
```
Comment (kytrinyx) [inline exercises/twelve-days/example.go]: What is a `frm`?

File: exercises/twelve-days/twelve_days_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+)
+
+const targetTestVersion = 1
+
+type testCase struct {
+	input    int
+	expected string
+}
+
+var singTestCases = []testCase{
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: We only have one set of test cases, so the variable can be `testCases` rather than `singTestCases`.

File: exercises/twelve-days/twelve_days_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	{10, "On the tenth day of Christmas my true love gave to me, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree."},
+	{11, "On the eleventh day of Christmas my true love gave to me, eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree."},
+	{12, "On the twelfth day of Christmas my true love gave to me, twelve Drummers Drumming, eleven Pipers Piping, ten Lords-a-Leaping, nine Ladies Dancing, eight Maids-a-Milking, seven Swans-a-Swimming, six Geese-a-Laying, five Gold Rings, four Calling Birds, three French Hens, two Turtle Doves, and a Partridge in a Pear Tree."},
+}
+
+func TestTestVersion(t *testing.T) {
+	if testVersion != targetTestVersion {
+		t.Errorf("Found testVersion = %v, want %v.", testVersion, targetTestVersion)
+	}
+}
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: Let's put this test at the very end of the file. The tests will execute in parallel, but it would be nice if this isn't the first thing you see when you read the tests since it's book-keeping.

File: exercises/twelve-days/twelve_days_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+func TestSing(t *testing.T) {
+	var song = ""
+	for i := 1; i <= 12; i++ {
+		song = fmt.Sprintf("%s\n", verses(i))
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: This is overwriting `song` 12 times, right?

Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: `verses` should probably be exported.

File: exercises/twelve-days/twelve_days_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	}
+	if song != sing() {
+		t.Errorf("Twelve Days test, error occured in sing test")
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: This says that there's an error, but there's no error handling in this test. Would you rephrase this to clarify why the test is failing?

File: exercises/twelve-days/twelve_days_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		t.Errorf("Found testVersion = %v, want %v.", testVersion, targetTestVersion)
+	}
+}
+
+func TestSing(t *testing.T) {
+	var song = ""
+	for i := 1; i <= 12; i++ {
+		song = fmt.Sprintf("%s\n", verses(i))
+	}
+	if song != sing() {
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: `sing()` should probably be exported.

File: exercises/twelve-days/twelve_days_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+		t.Errorf("Twelve Days test, error occured in sing test")
+	}
+}
+
+func TestVerse(t *testing.T) {
+	for testCnt, test := range singTestCases {
+		actual := verses(testCnt + 1)
```
Comment (kytrinyx) [inline exercises/twelve-days/twelve_days_test.go]: We've got `test.input` which is the correct value. We should use that here instead of `testCnt`.

File: exercises/twelve-days/twelve_days_test.go (hunk 2 above)
Comment (Akasurde) [inline exercises/twelve-days/twelve_days_test.go]: OK. I will change it to `testCases`

File: exercises/twelve-days/twelve_days_test.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	for i := 1; i <= 12; i++ {
+		song = fmt.Sprintf("%s\n", verses(i))
+	}
+	if song != sing() {
+		t.Errorf("Twelve Days test, error occured in sing test")
+	}
+}
+
+func TestVerse(t *testing.T) {
+	for testCnt, test := range singTestCases {
```
Comment (Akasurde) [inline exercises/twelve-days/twelve_days_test.go]: OK.

File: exercises/twelve-days/example.go (hunk 1 above)
Comment (Akasurde) [inline exercises/twelve-days/example.go]: I know, it is bad mnemonic. I converted from to frm, my bad.

File: config.json [hunk 9]
```diff
@@ -13,6 +13,7 @@
     "acronym",
     "bob",
     "triangle",
+    "twelve-days",
```
Comment (Akasurde) [inline config.json]: Could you please explain this more ?

File: exercises/twelve-days/twelve_days_test.go (hunk 7 above)
Comment (Akasurde) [inline exercises/twelve-days/twelve_days_test.go]: OK

File: exercises/twelve-days/twelve_days_test.go (hunk 5 above)
Comment (Akasurde) [inline exercises/twelve-days/twelve_days_test.go]: @kytrinyx  I am not following you here. Can you explain more ?

File: exercises/twelve-days/twelve_days_test.go (hunk 4 above)
Comment (Akasurde) [inline exercises/twelve-days/twelve_days_test.go]: what do you suggests here ?

**Comment by petertseng** [comment]
My only questions are:

* whether we are sure we want to leave `sing` and `verses` unexported.
* should it be `verses(i int)` where the function name is plural but it only takes a single argument? It seems to make more sense to be called `verse` or `Verse`. depending on whether we decide to export

**Comment by alebaffa** [comment]
In the exercise [food-chain](https://github.com/exercism/xgo/blob/master/exercises/food-chain/example.go) it's Verse (singular and exported). 
About exported functions or not: as I said in PR #396 , all the other exercises in xgo are exported. So, I think we should continue on this line.

**Comment by petertseng** [comment]
Because of inactivity: Rebasing for config.json

Rename sing -> Song and verses -> Verse

The food-chain move is going into a separate commit as well, I think.

**Comment by petertseng** [comment]
Happy late Hacktoberfest to @Akasurde , and thanks for the work.
